pkg/registry/pod/etcd/etcd.go

```diff
@@ -203,14 +218,28 @@ func (r *BindingREST) assignPod(ctx api.Context, podID string, machine string) e
 		return boundPodList, nil
 	})
 	if err != nil {
-		// Put the pod's host back the way it was. This is a terrible hack, but
-		// can't really be helped, since there's not really a way to do atomic
-		// multi-object changes in etcd.
-		if _, err2 := r.setPodHostTo(ctx, podID, machine, ""); err2 != nil {
+		glog.V(4).Infof("failing pod %v because constraint checks failed", podID)
+		podKey, err2 := r.store.KeyFunc(ctx, podID)
+		if err2 == nil {
+			err2 = r.store.Helper.AtomicUpdate(podKey, &api.Pod{}, false, func(obj runtime.Object) (runtime.Object, error) {
+				pod, ok := obj.(*api.Pod)
+				if !ok {
+					return nil, fmt.Errorf("unexpected object: %#v", obj)
+				}
+				if pod.Status.Host != machine {
+					return nil, fmt.Errorf("pod %v is already assigned to host %v", pod.Name, pod.Status.Host)
+				}
+				pod.Status.Host = "" //TODO(jdef) get rid of this once BoundPods goes away
+				pod.Status.Phase = api.PodFailed
+				pod.Status.Message = "HostAssignmentFailed"
```

Comment:
Pod status message should be more informative, such as including the host. Search the code for examples.